ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/RaceConditionIT.scala

```diff
@@ -0,0 +1,393 @@
+
+  raceConditionTest(
+    "WWDoubleNonTransientCreate",
+    "Cannot concurrently create multiple non-transient contracts with the same key",
+  ) { implicit ec => ledger => alice =>
+    val Attempts = 5
+    val ExpectedNumberOfSuccessfulCreations = 1
+    Future
+      .traverse(1 to Attempts) { _ =>
+        ledger.create(alice, ContractWithKey(alice)).transform(Success(_))
+      }
+      .map { results =>
+        assertLength(
+          "Successful contract creations",
+          ExpectedNumberOfSuccessfulCreations,
+          results.filter(_.isSuccess),
+        )
+        ()
+      }
+  }
+
+  raceConditionTest(
+    "WWDoubleArchive",
+    "Cannot archive the same contract multiple times",
+  ) { implicit ec => ledger => alice =>
+    val Attempts = 5
+    val ExpectedNumberOfSuccessfulArchivals = 1
+    for {
+      contract <- ledger.create(alice, ContractWithKey(alice))
+      _ <- Future.traverse(1 to Attempts) { _ =>
+        ledger.exercise(alice, contract.exerciseContractWithKey_Archive).transform(Success(_))
+      }
+      transactions <- transactions(ledger, alice)
+    } yield {
+      import TransactionUtil._
+      assertLength(
+        "Successful contract archivals",
+        ExpectedNumberOfSuccessfulArchivals,
+        transactions.filter(isArchival),
+      )
+      ()
+    }
+  }
+
+  raceConditionTest(
+    "WWArchiveVsNonTransientCreate",
+    "Cannot create a contract with a key if that key is still used by another contract",
+  ) { implicit ec => ledger => alice =>
+    /*
+    This test case is intended to catch a race condition ending up in two consecutive successful contract
+    create or archive commands. E.g.:
+    [create]  <wait>  [archive]-race-[create]
+    In case of a bug causing the second [create] to see a partial result of [archive] command we could end up
+    with two consecutive successful contract creations.
+     */
+    for {
+      contract <- ledger.create(alice, ContractWithKey(alice))
+      _ <- Delayed.by(1.second)(())
+      createFuture = ledger.create(alice, ContractWithKey(alice)).transform(Success(_))
+      exerciseFuture = ledger
+        .exercise(alice, contract.exerciseContractWithKey_Archive)
+        .transform(Success(_))
+      _ <- createFuture
+      _ <- exerciseFuture
+      _ <- ledger.create(
+        alice,
+        DummyContract(alice),
+      ) // Create a dummy contract to ensure that we're not stuck with previous commands
+      transactions <- transactions(ledger, alice)
+    } yield {
+      import TransactionUtil._
+
+      assert(
+        isCreateNonTransient(transactions.head),
+        "The first transaction is expected to be a contract creation",
+      )
+      assert(
+        transactions.exists(isCreateDummyContract),
+        "A dummy contract creation is missing. A possible reason might be reading the transactions stream in the test case before submitted commands had chance to be processed.",
+      )
+
+      val (_, valid) =
```

Comment:
Can't we make a more strict assertion here? Don't we know how many transactions should there be and their order?

Reply:
Actually after some discussions we came with a conclusion that we only care about the order here and not how many transactions (because of contention that might happen here). That is why this simple check is sufficient for our needs.